src/Application/UseCases/GetPagesWithoutPersistentIdentifiers.php

```diff
@@ -0,0 +1,28 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace ProfessionalWiki\PersistentPageIdentifiers\Application\UseCases;
+
+use Wikimedia\Rdbms\IDatabase;
+
+class GetPagesWithoutPersistentIdentifiers {
+
+	public function __construct(
+		private readonly IDatabase $database
+	) {
+	}
+
+	/**
+	 * @return string[] page IDs
+	 */
+	public function getPagesWithoutPersistentIdentifiers(): array {
+		return $this->database->newSelectQueryBuilder()
+			->select( 'p.page_id' )
+			->from( 'page', 'p' )
+			->leftJoin( 'persistent_page_ids', 'ppi', 'p.page_id = ppi.page_id' )
+			->where( 'ppi.page_id IS NULL' )
+			->fetchFieldValues();
```

Comment:
This works for all non-deleted pages.
If we need to generate persistent IDs for deleted pages then we'll need to do a similar query, but using field `ar_page_id` on the `archive` table.